temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java

```diff
@@ -132,16 +124,27 @@ class ReplayWorkflowRunTaskHandler implements WorkflowRunTaskHandler {
   }
 
   @Override
-  public WorkflowTaskResult handleWorkflowTask(PollWorkflowTaskQueueResponseOrBuilder workflowTask)
+  public WorkflowTaskResult handleWorkflowTask(
+      PollWorkflowTaskQueueResponseOrBuilder workflowTask, WorkflowHistoryIterator historyIterator)
       throws InterruptedException {
     lock.lock();
     try {
       long startTimeNanos = System.nanoTime();
-      handleWorkflowTaskImpl(workflowTask);
+
+      if (workflowTask.getPreviousStartedEventId()
+          < workflowStateMachines.getCurrentStartedEventId()) {
+        // if previousStartedEventId < currentStartedEventId - the last workflow task handled by
+        // these state machines is ahead of the last handled workflow task known by the server.
+        // Something is off, the server lost progress.
+        throw new IllegalStateException(
+            "Server history for the workflow is below the progress of the workflow on the worker, the progress needs to be discarded");
```

Comment:
Technically, you can blow the cache and fetch history here instead. This is so rare though, it seems fine to not bother with that code.

Reply:
I can, but it likely doesn't make much sense if it's actually an outdated WFT for some reason. I will leave a comment here that it should be done if it's ever become any common occurrence or a problem for anything.